www/screen_shot.php

```diff
@@ -230,6 +231,58 @@ function printStep($fileHandler, $testInfo, $testStepResult, $useQuicklinks) {
         echo '<img class="center" alt="Responsive Site Check image" src="' . $urlPaths->additionalScreenShotFile("responsive") . '">';
     }
 
+    // Generate and display hero element screen shots
+    if (gz_is_file($localPaths->heroElementsJsonFile())) {
+        $hero_data = json_decode(gz_file_get_contents($localPaths->heroElementsJsonFile()), true);
+        $visual_progress = $testStepResult->getVisualProgress();
+
+        if (isset($visual_progress['frames']) && $hero_data) {
+            $frames = $visual_progress['frames'];
+            $hero_elements = $hero_data['heroes'];
+            $hero_times = $testStepResult->getMetric('heroElementTimes');
+            $viewport = $hero_data['viewport'];
+
+            foreach ($hero_elements as $hero) {
+                $hero_time = $hero_times[$hero['name']];
+
+                if ($hero_time > 0 && isset($frames[$hero_time])) {
+                    $frame_path = $frames[$hero_time]['path'];
+                    if (strtolower(substr($frame_path, -4)) == '.png') {
+                        $frame = imagecreatefrompng("./{$frame_path}");
+                    } else {
+                        $frame = imagecreatefromjpeg("./{$frame_path}");
+                    }
+
+                    $frame_w = imagesx($frame);
+                    $frame_h = imagesy($frame);
+                    $scale = min($frame_w / $viewport['width'], $frame_h / $viewport['height']);
+                    $x = $hero['x'] * $scale;
+                    $y = $hero['y'] * $scale;
+                    $w = $hero['width'] * $scale;
+                    $h = $hero['height'] * $scale;
+
+                    // Take a copy of the frame for the "highlight" image
```

Comment:
Looks like this re-generates the file every time the screenshot page is accessed.  At a minimum it should check to see if the file already exists first.

That said, is there any way to do the highlighting entirely on the DOM by overlaying DIV's over the image with different transparency (and a border for the green highlight)?  That would reduce the CPU hit and storage space on the server generating images for each hero.  Probably not a big deal since the screenshot page isn't accessed that much but would like to see if it's possible first.